---
version: '3.8'

x-logging: &default-logging
  driver: "json-file"
  options:
    max-file: "5"
    max-size: "10m"

x-healthcheck: &default-healthcheck
  test: ["CMD", "nc", "-z", "localhost", "port"]
  timeout: 45s
  interval: 10s
  retries: 10

x-environment: &default-environment
  CLUSTER_NAME: hadoop_cluster

x-restart: &default-restart
  restart: always

services:
  # Serviço do Namenode
  infra-namenode:
    image: iamgacarvalho/hadoop-namenode-data-in-compass:0.0.3
    container_name: namenode
    networks:
      - hadoop_network
    logging: *default-logging
    ports:
      - "32763:9870"
    volumes:
      - ./mnt/hadoop/namenode:/hadoop/dfs/name
    environment: *default-environment
    healthcheck:
      <<: *default-healthcheck
      test: ["CMD", "nc", "-z", "namenode", "9870"]

  # Serviço do Datanode
  infra-datanode:
    image: iamgacarvalho/hadoop-datanode-data-in-compass:0.0.3
    container_name: datanode
    networks:
      - hadoop_network
    logging: *default-logging
    volumes:
      - ./mnt/hadoop/datanode:/hadoop/dfs/data
    environment:
      SERVICE_PRECONDITION: namenode:9870
      DATANODE_PORT: 9864
    healthcheck:
      <<: *default-healthcheck
      test: ["CMD", "nc", "-z", "datanode", "9864"]
    depends_on:
      - infra-namenode
    deploy:
      replicas: 3

volumes:
  infra-namenode:
  infra-datanode:

networks:
  hadoop_network:
    external: true
